src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -3137,5 +3132,346 @@ static void M()
+        ConstantFunc();
+
+        const int ConstantExpr1 = 1;
+        const int ConstantExpr2 = 4;
+        void ConstantExprFunc(int x = ConstantExpr1 + ConstantExpr2) => System.Console.WriteLine(x);
+        ConstantExprFunc();
+
+        void StrAdd1() { }
+        void StrAdd2() { }
+        void StrAddFunc(string s = nameof(StrAdd1) + nameof(StrAdd2)) => System.Console.WriteLine(s);
+        StrAddFunc();
+
+        void ClassConstantFunc(int x = ClassConstant * 1) => System.Console.WriteLine(x);
+        ClassConstantFunc();
+
+        const bool ConstBoolFalse = false;
+        const bool ConstBoolTrue = true;
+        void LogicalExpr(bool x = ConstBoolFalse || ConstBoolTrue) => System.Console.WriteLine(x);
+        LogicalExpr();
+
+        void GenericLocal<T>() => System.Console.WriteLine(typeof(T).Name);
+
+        void GenericLocalFunc(string s = nameof(GenericLocal)) => System.Console.WriteLine(s);
+        GenericLocalFunc();
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: DefaultParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(16821, "https://github.com/dotnet/roslyn/issues/16821")]
+        public void MethodReportsDefaultParametersAsUsed()
+        {
+            var source = @"
+using System;
+internal abstract class Abstract
+{
+    static int FuncVar = 2;
+    protected abstract void Func(string a = nameof(FuncVar));
+}
+class AttributeArgument : Attribute
+{
+    static int Field = 0;
+    public string Instance;
+    [AttributeArgument(nameof(Field))]
+    public AttributeArgument(string param)
+    { }
+}
+class AttributeNamedArgument : Attribute
+{
+    static int Field = 0;
+    public string Instance;
+    [AttributeNamedArgument("""", Instance = nameof(Field))]
+    AttributeNamedArgument(string param)
+    { }
+}
+class GenericClass<T>
+{
+    public static T Field = default(T);
+}
+class Program
+{
+    static void RefGenericField(string param1 = nameof(GenericClass<int>.Field))
+    { }
+
+    static int F1var = 2;
+    static void F1(string a = nameof(F1var)) { }
+
+    static int F2var = 2;
+    int this[int dummy, string a = nameof(F2var)] { set { } }
+
+    static int F3var = 2;
+    static void AttributeInParam([AttributeArgument(nameof(F3var))] string s) { }
+    static int F4var = 2;
+    static void NamedAttributeInParam([AttributeArgument(param: nameof(F4var))] string s) { }
+    static int F5var = 2;
+    static void NamedAttributeInParam2([AttributeArgument(""2"", Instance = nameof(F5var))] string s) { }
+
+    static void Main()
+    {
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: DefaultParseOptions, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(16821, "https://github.com/dotnet/roslyn/issues/16821")]
+        public void SpeculativeDoesNotMarkUsed()
```

Comment:
>SpeculativeDoesNotMarkUsed [](start = 20, length = 26)

We should have similar speculation tests for default parameters and for local functions, including speculation on local function inside a lambda expressions.